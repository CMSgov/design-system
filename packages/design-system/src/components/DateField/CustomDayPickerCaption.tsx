import type * as React from 'react';
import { ArrowIcon } from '../Icons';
import { CaptionDropdowns, CaptionProps, useDayPicker, useNavigation } from 'react-day-picker';

/**
 * Implements custom month/year navigation controls and caption for the DayPicker
 * component. Much of this code is copied out of the `react-day-picker` source and
 * then modified to fit our needs.
 */
export function CustomDayPickerCaption(props: CaptionProps) {
  const {
    classNames,
    styles,
    labels: { labelPrevious, labelNext },
    locale,
    onMonthChange,
  } = useDayPicker();
  const { previousMonth, nextMonth, goToMonth } = useNavigation();

  const previousLabel = labelPrevious(previousMonth, { locale });
  const nextLabel = labelNext(nextMonth, { locale });

  const handlePreviousClick: React.MouseEventHandler = () => {
    if (!previousMonth) return;
    goToMonth(previousMonth);
    onMonthChange?.(previousMonth);
  };

  const handleNextClick: React.MouseEventHandler = () => {
    if (!nextMonth) return;
    goToMonth(nextMonth);
    onMonthChange?.(nextMonth);
  };

  return (
    <div className={classNames.caption} style={styles.caption}>
      <button
        aria-label={previousLabel}
        className="ds-c-single-input-date-field__nav"
        onClick={handlePreviousClick}
        type="button"
        disabled={!previousMonth}
      >
        <ArrowIcon direction="left" />
      </button>
      <CaptionDropdowns displayMonth={props.displayMonth} id={props.id} />
      <button
        aria-label={nextLabel}
        className="ds-c-single-input-date-field__nav"
        onClick={handleNextClick}
        type="button"
        disabled={!nextMonth}
      >
        <ArrowIcon direction="right" />
      </button>
    </div>
  );
}

export default CustomDayPickerCaption;
