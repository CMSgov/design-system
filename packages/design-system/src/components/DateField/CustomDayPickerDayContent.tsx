import { ActiveModifiers, useDayPicker } from 'react-day-picker';

/** Represent the props for the {@link DayContent} component. */
interface DayContentProps {
  /** The date representing the day. */
  date: Date;
  /** The month where the day is displayed. */
  displayMonth: Date;
  /** The active modifiers for the given date. */
  activeModifiers: ActiveModifiers;
}

/**
 * Render the content of the day cell.
 */
/** Render the content of the day cell. */
export function CustomDayPickerDayContent(props: DayContentProps) {
  const {
    locale,
    classNames,
    styles,
    labels: { labelDay },
    formatters: { formatDay },
  } = useDayPicker();

  return (
    <>
      <span aria-hidden="true">{formatDay(props.date, { locale })}</span>
      <span className={classNames.vhidden} style={styles.vhidden}>
        {labelDay(props.date, props.activeModifiers, { locale })}
      </span>
    </>
  );
}
