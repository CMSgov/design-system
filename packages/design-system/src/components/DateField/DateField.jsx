import FormLabel from '../FormLabel/FormLabel';
import PropTypes from 'prop-types';
import React from 'react';
import TextField from '../TextField/TextField';
import classNames from 'classnames';
import uniqueId from 'lodash.uniqueid';

// Prevents day/month greater than 2 digits and year greater than 4 digits
const standardLengthFormatter = ({ day, month, year }) => ({
  day: day.length > 2 ? day.substring(0, 2) : day,
  month: month.length > 2 ? month.substring(0, 2) : month,
  year: year.length > 4 ? year.substring(0, 4) : year
});

export const defaultDateFormatter = dateObject => {
  const standardDate = standardLengthFormatter(dateObject);
  return standardDate;
};

export class DateField extends React.PureComponent {
  constructor(props) {
    super(props);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  labelId() {
    if (!this._labelId) {
      // Cache the ID so we're not regenerating it on each method call
      this._labelId = uniqueId('datefield_label_');
    }

    return this._labelId;
  }

  formatDate() {
    if (this.props.dateFormatter && this.monthInput && this.dayInput && this.yearInput) {
      const values = {
        month: this.monthInput.value,
        day: this.dayInput.value,
        year: this.yearInput.value
      };

      return this.props.dateFormatter(values);
    }
  }

  handleBlur(evt) {
    if (this.props.onBlur) {
      this.props.onBlur(evt, this.formatDate());
    }

    if (this.props.onComponentBlur) {
      this.handleComponentBlur(evt);
    }
  }

  handleChange(evt) {
    this.props.onChange(evt, this.formatDate());
  }

  handleComponentBlur(evt) {
    // The active element is always the document body during a focus
    // transition, so in order to check if the newly focused element
    // is one of our other date inputs, we're going to have to wait
    // a bit.
    setTimeout(() => {
      if (
        document.activeElement !== this.dayInput &&
        document.activeElement !== this.monthInput &&
        document.activeElement !== this.yearInput
      ) {
        this.props.onComponentBlur(evt, this.formatDate());
      }
    }, 20);
  }

  render() {
    const sharedDateFieldProps = {
      className: 'ds-l-col--auto',
      labelClassName: 'ds-u-font-weight--normal ds-u-margin-top--1',
      inversed: this.props.inversed,
      onBlur: (this.props.onBlur || this.props.onComponentBlur) && this.handleBlur,
      onChange: this.props.onChange && this.handleChange,
      numeric: true
    };
    const labelId = this.labelId();

    return (
      <fieldset className="ds-c-fieldset">
        <FormLabel
          component="legend"
          errorMessage={this.props.errorMessage}
          hint={this.props.hint}
          inversed={this.props.inversed}
          requirementLabel={this.props.requirementLabel}
          id={labelId}
        >
          {this.props.label}
        </FormLabel>

        <div className="ds-l-form-row ds-u-align-items--end">
          <TextField
            {...sharedDateFieldProps}
            fieldClassName={classNames('ds-c-field--month', {
              'ds-c-field--error': this.props.monthInvalid
            })}
            inputRef={el => {
              this.monthInput = el;
              if (this.props.monthFieldRef) this.props.monthFieldRef(el);
            }}
            defaultValue={this.props.monthDefaultValue}
            label={this.props.monthLabel}
            name={this.props.monthName}
            value={this.props.monthValue}
            aria-describedby={labelId}
          />
          <span className="ds-c-datefield__separator">/</span>
          <TextField
            {...sharedDateFieldProps}
            fieldClassName={classNames('ds-c-field--day', {
              'ds-c-field--error': this.props.dayInvalid
            })}
            inputRef={el => {
              this.dayInput = el;
              if (this.props.dayFieldRef) this.props.dayFieldRef(el);
            }}
            defaultValue={this.props.dayDefaultValue}
            label={this.props.dayLabel}
            name={this.props.dayName}
            value={this.props.dayValue}
            aria-describedby={labelId}
          />
          <span className="ds-c-datefield__separator">/</span>
          <TextField
            {...sharedDateFieldProps}
            fieldClassName={classNames('ds-c-field--year', {
              'ds-c-field--error': this.props.yearInvalid
            })}
            inputRef={el => {
              this.yearInput = el;
              if (this.props.yearFieldRef) this.props.yearFieldRef(el);
            }}
            defaultValue={this.props.yearDefaultValue}
            label={this.props.yearLabel}
            name={this.props.yearName}
            value={this.props.yearValue}
            aria-describedby={labelId}
          />
        </div>
      </fieldset>
    );
  }
}

DateField.defaultProps = {
  label: 'Date',
  hint: 'For example: 4 / 28 / 1986',
  dayLabel: 'Day',
  dayName: 'day',
  monthLabel: 'Month',
  monthName: 'month',
  yearLabel: 'Year',
  yearName: 'year',
  dateFormatter: defaultDateFormatter
};

DateField.propTypes = {
  /**
   * Optional method to format the `input` field values. If this
   * method is provided, the returned value will be passed as a second argument
   * to the `onBlur` and `onChange` callbacks. This method receives an object as
   * its only argument, in the shape of: `{ day, month, year }`
   *
   * By default `dateFormatter` will be set to the `defaultDateFormatter` function, which prevents days/months more than 2 digits & years more than 4 digits.
   */
  dateFormatter: PropTypes.func,
  errorMessage: PropTypes.node,
  /**
   * Additional hint text to display above the individual month/day/year fields
   */
  hint: PropTypes.node,
  /**
   * Applies the "inverse" UI theme
   */
  inversed: PropTypes.bool,
  /**
   * The primary label, rendered above the individual month/day/year fields
   */
  label: PropTypes.node,
  /**
   * Text showing the requirement ("Required", "Optional", etc.). See [Required and Optional Fields]({{root}}/guidelines/forms/#required-and-optional-fields).
   */
  requirementLabel: PropTypes.node,
  /**
   * Called anytime any date input is blurred
   */
  onBlur: PropTypes.func,
  /**
   * Called when any date input is blurred and the focus does not land on one
   * of the other date inputs inside this component (i.e., when the whole
   * component loses focus)
   */
  onComponentBlur: PropTypes.func,
  /**
   * Called anytime any date input is changed
   */
  onChange: PropTypes.func,
  /**
   * Label for the day field
   */
  dayLabel: PropTypes.node,
  /**
   * `name` for the day `input` field
   */
  dayName: PropTypes.string,
  /**
   * Initial value for the day `input` field. Use this for an uncontrolled
   * component; otherwise, use the `dayValue` property.
   */
  dayDefaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Access a reference to the day `input`
   */
  dayFieldRef: PropTypes.func,
  /**
   * Apply error styling to the day `input`
   */
  dayInvalid: PropTypes.bool,
  /**
   * Sets the day input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `dayDefaultValue`.
   */
  dayValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Label for the month field
   */
  monthLabel: PropTypes.node,
  /**
   * `name` for the month `input` field
   */
  monthName: PropTypes.string,
  /**
   * Initial value for the month `input` field. Use this for an uncontrolled
   * component; otherwise, use the `monthValue` property.
   */
  monthDefaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Access a reference to the month `input`
   */
  monthFieldRef: PropTypes.func,
  /**
   * Apply error styling to the month `input`
   */
  monthInvalid: PropTypes.bool,
  /**
   * Sets the month input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `monthDefaultValue`.
   */
  monthValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Initial value for the year `input` field. Use this for an uncontrolled
   * component; otherwise, use the `yearValue` property.
   */
  yearDefaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Access a reference to the year `input`
   */
  yearFieldRef: PropTypes.func,
  /**
   * Apply error styling to the year `input`
   */
  yearInvalid: PropTypes.bool,
  /**
   * Label for the year `input` field
   */
  yearLabel: PropTypes.node,
  /**
   * `name` for the year field
   */
  yearName: PropTypes.string,
  /**
   * Sets the year input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `yearDefaultValue`.
   */
  yearValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

export default DateField;
