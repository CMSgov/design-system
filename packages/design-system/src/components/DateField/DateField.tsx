import { FormControl, FormControlPropKeys } from '../FormControl/FormControl';
import DateInput from './DateInput';
import React from 'react';
import defaultDateFormatter from './defaultDateFormatter';
import omit from 'lodash/omit';
import pick from 'lodash/pick';

export type DateFieldDayDefaultValue = string | number;
export type DateFieldDayValue = string | number;
export type DateFieldMonthDefaultValue = string | number;
export type DateFieldMonthValue = string | number;
export type DateFieldYearDefaultValue = string | number;
export type DateFieldYearValue = string | number;

export interface DateFieldProps {
  /**
   * Adds `autocomplete` attributes `bday-day`, `bday-month` and `bday-year` to the corresponding `<DateField>` inputs
   */
  autoComplete?: boolean;
  /**
   * Additional classes to be added to the root fieldset element
   */
  className?: string;
  /**
   * Optional method to format the `input` field values. If this
   * method is provided, the returned value will be passed as a second argument
   * to the `onBlur` and `onChange` callbacks. This method receives an object as
   * its only argument, in the shape of: `{ day, month, year }`
   *
   * By default `dateFormatter` will be set to the `defaultDateFormatter` function, which prevents days/months more than 2 digits & years more than 4 digits.
   */
  dateFormatter?: typeof defaultDateFormatter;
  disabled?: boolean;
  errorMessage?: React.ReactNode;
  /**
   * Additional classes to be added to the error message
   */
  errorMessageClassName?: string;
  /**
   * Location of the error message relative to the field input
   */
  errorPlacement?: 'top' | 'bottom';
  /**
   * Additional hint text to display above the individual month/day/year fields
   */
  hint?: React.ReactNode;
  /**
   * Applies the "inverse" UI theme
   */
  inversed?: boolean;
  /**
   * The primary label, rendered above the individual month/day/year fields
   */
  label?: React.ReactNode;
  /**
   * A unique ID to be used for the DateField label. If one isn't provided, a unique ID will be generated.
   */
  labelId?: string;
  /**
   * Text showing the requirement ("Required", "Optional", etc.). See [Required and Optional Fields]({{root}}/guidelines/forms/#required-and-optional-fields).
   */
  requirementLabel?: React.ReactNode;
  /**
   * Called anytime any date input is blurred
   */
  onBlur?: (...args: any[]) => any;
  /**
   * Called when any date input is blurred and the focus does not land on one
   * of the other date inputs inside this component (i.e., when the whole
   * component loses focus)
   */
  onComponentBlur?: (...args: any[]) => any;
  /**
   * Called anytime any date input is changed
   */
  onChange?: (...args: any[]) => any;
  /**
   * Label for the day field
   */
  dayLabel?: React.ReactNode;
  /**
   * `name` for the day `input` field
   */
  dayName?: string;
  /**
   * Initial value for the day `input` field. Use this for an uncontrolled
   * component; otherwise, use the `dayValue` property.
   */
  dayDefaultValue?: DateFieldDayDefaultValue;
  /**
   * Access a reference to the day `input`
   */
  dayFieldRef?: (...args: any[]) => any;
  /**
   * Apply error styling to the day `input`
   */
  dayInvalid?: boolean;
  /**
   * Sets the day input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `dayDefaultValue`.
   */
  dayValue?: DateFieldDayValue;
  /**
   * Label for the month field
   */
  monthLabel?: React.ReactNode;
  /**
   * `name` for the month `input` field
   */
  monthName?: string;
  /**
   * Initial value for the month `input` field. Use this for an uncontrolled
   * component; otherwise, use the `monthValue` property.
   */
  monthDefaultValue?: DateFieldMonthDefaultValue;
  /**
   * Access a reference to the month `input`
   */
  monthFieldRef?: (...args: any[]) => any;
  /**
   * Apply error styling to the month `input`
   */
  monthInvalid?: boolean;
  /**
   * Sets the month input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `monthDefaultValue`.
   */
  monthValue?: DateFieldMonthValue;
  /**
   * Initial value for the year `input` field. Use this for an uncontrolled
   * component; otherwise, use the `yearValue` property.
   */
  yearDefaultValue?: DateFieldYearDefaultValue;
  /**
   * Access a reference to the year `input`
   */
  yearFieldRef?: (...args: any[]) => any;
  /**
   * Apply error styling to the year `input`
   */
  yearInvalid?: boolean;
  /**
   * Label for the year `input` field
   */
  yearLabel?: React.ReactNode;
  /**
   * `name` for the year field
   */
  yearName?: string;
  /**
   * Sets the year input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `yearDefaultValue`.
   */
  yearValue?: DateFieldYearValue;
}

export function DateField(props: DateFieldProps): React.ReactElement {
  const containerProps = pick(props, FormControlPropKeys);
  const inputOnlyProps = omit(props, FormControlPropKeys);

  return (
    <FormControl
      {...containerProps}
      component="fieldset"
      labelComponent="legend"
      render={({ labelId }) => (
        <DateInput {...inputOnlyProps} {...{ labelId }} inversed={props.inversed} />
      )}
    />
  );
}

DateField.defaultProps = {
  label: 'Date',
  hint: 'For example: 4 / 28 / 1986',
  dayLabel: 'Day',
  dayName: 'day',
  monthLabel: 'Month',
  monthName: 'month',
  yearLabel: 'Year',
  yearName: 'year',
  dateFormatter: defaultDateFormatter,
};

export default DateField;
