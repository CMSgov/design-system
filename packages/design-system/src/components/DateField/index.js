export { default as DateField } from './DateField';
export { default as DateInput } from './DateInput';
