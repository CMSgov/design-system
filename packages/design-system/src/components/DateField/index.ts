export { default as DateField } from './MultiInputDateField';
export { default as SingleInputDateField } from './SingleInputDateField';
export { default as DateInput } from './DateInput';
