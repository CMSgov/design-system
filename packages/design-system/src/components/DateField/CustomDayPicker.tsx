import React from 'react';
import CustomDayPickerCaption from './CustomDayPickerCaption';
import { format } from 'date-fns';
import { DayPicker, DayPickerSingleProps } from 'react-day-picker';
import { ArrowsStackedIcon } from '../Icons';
import { t } from '../i18n';
import type { Locale } from 'date-fns';

/**
 * The default formatter for the Month caption.
 */
export function formatMonthCaption(month: Date, options?: { locale?: Locale }) {
  return format(month, 'LLL', options);
}

type CustomDayPickerProps = Pick<
  DayPickerSingleProps,
  | 'selected'
  | 'onSelect'
  | 'defaultMonth'
  | 'fromDate'
  | 'fromMonth'
  | 'fromYear'
  | 'toDate'
  | 'toMonth'
  | 'toYear'
>;

export function CustomDayPicker(props: CustomDayPickerProps) {
  return (
    <DayPicker
      mode="single"
      footer={
        <span className="ds-u-visibility--screen-reader">
          {t('singleInputDateField.arrowKeyInstructions')}
        </span>
      }
      captionLayout="dropdown"
      components={{ Caption: CustomDayPickerCaption, IconDropdown: ArrowsStackedIcon }}
      formatters={{ formatMonthCaption }}
      {...props}
    />
  );
}

export default CustomDayPicker;
