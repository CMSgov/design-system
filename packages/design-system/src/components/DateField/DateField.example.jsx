import React, { Fragment } from 'react';
import DateField from './DateField';
import ReactDOM from 'react-dom';

class ControlledDateField extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      day: '10',
      month: '1',
      year: '2000'
    };
  }

  render() {
    return (
      <fieldset className="ds-u-margin-top--3 ds-u-padding-left--3 ds-u-padding-bottom--3 ds-u-border--2">
        <legend className="ds-u-font-size--h3">Controlled Example</legend>
        <DateField
          label={
            <span>
              DateField using default <code>dateFormatter</code>
            </span>
          }
          hint={'Try to enter a date with invalid number of digits'}
          monthValue={this.state.month}
          dayValue={this.state.day}
          yearValue={this.state.year}
          onChange={(e, dateObject) => this.setState(dateObject)}
        />
      </fieldset>
    );
  }
}

ReactDOM.render(
  <Fragment>
    <DateField
      label="DateField example"
      errorMessage="Please enter a year in the past"
      monthDefaultValue="10"
      dayDefaultValue="31"
      yearDefaultValue="2020"
      yearInvalid
    />
    <ControlledDateField />
  </Fragment>,
  document.getElementById('js-example')
);
