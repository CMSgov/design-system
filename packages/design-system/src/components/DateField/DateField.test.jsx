jest.mock('lodash.uniqueid', () => (str) => `${str}snapshot`);
/* eslint-disable import/first */
import { DateField, defaultDateFormatter } from './DateField';
import React from 'react';
import renderer from 'react-test-renderer';

describe('DateField', () => {
  it('renders', () => {
    expect(renderer.create(<DateField />)).toMatchSnapshot();
  });

  describe('defaultDateFormatter', () => {
    it('prevents day over 2 digits', () => {
      const date = { day: '100', month: '10', year: '2000' };
      const formattedDate = defaultDateFormatter(date);
      expect(formattedDate.day).toBe('10');
      expect(formattedDate.month).toBe(date.month);
      expect(formattedDate.year).toBe(date.year);
    });

    it('prevents month over 2 digits', () => {
      const date = { day: '10', month: '100', year: '2000' };
      const formattedDate = defaultDateFormatter(date);
      expect(formattedDate.day).toBe(date.day);
      expect(formattedDate.month).toBe('10');
      expect(formattedDate.year).toBe(date.year);
    });

    it('prevents year over 4 digits', () => {
      const date = { day: '10', month: '10', year: '20000' };
      const formattedDate = defaultDateFormatter(date);
      expect(formattedDate.day).toBe(date.day);
      expect(formattedDate.month).toBe(date.month);
      expect(formattedDate.year).toBe('2000');
    });
  });
});
