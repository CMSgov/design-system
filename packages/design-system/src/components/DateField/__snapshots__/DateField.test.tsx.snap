// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateField renders 1`] = `
<fieldset
  className="ds-c-fieldset"
>
  <legend
    className="ds-c-label"
    id="field_snapshot-label"
  >
    <span
      className=""
    >
      Date
    </span>
    <span
      className="ds-c-field__hint"
    >
      For example: 4 / 28 / 1986
    </span>
  </legend>
  <div
    className="ds-c-datefield__container ds-l-form-row"
  >
    <div
      className="ds-u-clearfix ds-l-col--auto"
    >
      <label
        className="ds-c-label ds-c-datefield__label"
        htmlFor="field_snapshot"
        id="field_snapshot-label"
      >
        <span
          className=""
        >
          Month
        </span>
      </label>
      <input
        aria-describedby="field_snapshot-label"
        className="ds-c-field ds-c-field--month"
        id="field_snapshot"
        inputMode="numeric"
        name="month"
        pattern="[0-9]*"
        type="text"
      />
    </div>
    <span
      className="ds-c-datefield__separator"
    >
      /
    </span>
    <div
      className="ds-u-clearfix ds-l-col--auto"
    >
      <label
        className="ds-c-label ds-c-datefield__label"
        htmlFor="field_snapshot"
        id="field_snapshot-label"
      >
        <span
          className=""
        >
          Day
        </span>
      </label>
      <input
        aria-describedby="field_snapshot-label"
        className="ds-c-field ds-c-field--day"
        id="field_snapshot"
        inputMode="numeric"
        name="day"
        pattern="[0-9]*"
        type="text"
      />
    </div>
    <span
      className="ds-c-datefield__separator"
    >
      /
    </span>
    <div
      className="ds-u-clearfix ds-l-col--auto"
    >
      <label
        className="ds-c-label ds-c-datefield__label"
        htmlFor="field_snapshot"
        id="field_snapshot-label"
      >
        <span
          className=""
        >
          Year
        </span>
      </label>
      <input
        aria-describedby="field_snapshot-label"
        className="ds-c-field ds-c-field--year"
        id="field_snapshot"
        inputMode="numeric"
        name="year"
        pattern="[0-9]*"
        type="text"
      />
    </div>
  </div>
</fieldset>
`;
